# 2.3 Arithmetik
Viele Programme führen arithmetische Berechnungen durch. Die folgende Tabelle fasst die arithmetischen Operatoren zusammen, die einige in der Algebra nicht verwendete Symbole enthalten.

![Arithmetic Operators](ch02images/arithmetic_operators.png)



## Multiplikation (*)

In [1]:
7 * 4

28

## Potenz (**)
Der Potenzierungsoperator (**) erhöht einen Wert um die Potenz eines anderen:

In [2]:
2 ** 3

8

Um die Quadratwurzel zu berechnen, können Sie den Exponenten 1/2 (d.h. 0.5) verwenden:

In [3]:
9 ** (1/2)

3.0

In [4]:
9 ** 0.5

3.0

## Division (/)
Die Division (/) dividiert einen Zähler durch einen Nenner und ergibt eine Fliesskommazahl mit einem Dezimalpunkt.

In [5]:
7 / 4

1.75

In [6]:
-13 / 4

-3.25

##  Ganzzahl-Division (//)
Die Ganzahhl-Division (//) ist eine Division ohne Rest.
Die Ganzzahl-Division dividiert einen Zähler durch einen Nenner, und gibt als Resultat die höchste _ganze_ Zahl zurück, die nicht grösser als das Ergebnis der echten Division ist. D.h., Python verwirft den Bruchteil.

In [7]:
7 // 4

1

In [8]:
3 // 5

0

In [9]:
14 // 7

2

In [10]:
-13 // 4

-4

Die Ganzzahl-Division ergibt die nächste ganze Zahl zurück, die nicht grösser als -3.25 ist, also -4.

## Ausnahmen (Exceptions)

### ZeroDivisionError
Die Division durch Null mit / oder // ist nicht erlaubt und führt zu einer Ausnahme (Exception) - ein Zeichen dafür, dass ein Problem aufgetreten ist:

In [11]:
123 / 0

ZeroDivisionError: division by zero

In [3]:
try: 
    123 / 0
except ZeroDivisionError:
    print('not possible to divide by zero')

anyexepoccured


Python meldet eine Ausnahme mit einem **Traceback**. Der Traceback in unserem Beispiel zeigt an, dass eine Ausnahme vom Typ _ZeroDivisionError_ aufgetreten ist.

Im interaktiven Modus wird die Snippet-Nummer, die die Ausnahme verursacht hat durch `Cell In [11]` angegeben. 

Die Zeile, die mit ----> 1 beginnt, zeigt den Code, der die Ausnahme verursacht hat.
Oft haben Snippets mehr als eine Codezeile - die 1 rechts von ----> zeigt an, dass die erste Codezeile innerhalb des Snippets die Ausnahme verursacht hat.

Die letzte Zeile des Traceback zeigt die aufgetretene Ausnahme, gefolgt von einem Doppelpunkt (:) und einer Fehlermeldung mit weiteren Informationen über die Ausnahme an.

### NameError
Eine Ausnahme tritt auch dann auf, wenn Sie versuchen, eine Variable zu verwenden, die Sie noch nicht erstellt haben. Das folgende Snippet versucht, der undefinierten Variable `z` den Wert 7 hinzuzufügen, was zu einem NameError führt:

In [12]:
z + 7

NameError: name 'z' is not defined

## Modulo (Restwert) Operator (%)
Der Python-Modulo-Operator (%) berechnet den Restwert, nachdem der linke Operand durch den rechten Operanden dividiert wurde:

In [1]:
17 % 5

2

In diesem Fall ergibt 17 geteilt durch 5 einen Quotienten von 3 und einen Rest von 2.

Dieser Operator wird am häufigsten bei ganzen Zahlen verwendet, kann aber auch mit anderen numerischen Typen verwendet werden:

In [16]:
7.5 % 3.5

0.5

## Gruppierung von Ausdrücken mit Klammern
Klammern gruppieren Python-Ausdrücke, analog wie sie in algebraischen Ausdrücken vorkommen. Zum Beispiel multipliziert der folgende Code das Resultat von 5 + 3 mit 10:

In [17]:
10 * (5 + 3)

80

Ohne diese Klammern ist das Ergebnis anders:

In [18]:
10 * 5 + 3

53

## Vorrangregeln von Operatoren
#### 1. Ausdrücke in Klammern werden zuerst ausgewertet
* Klammern haben die höchste Priorität. Sie können die Reihenfolge der Anweisung erzwingen
* Bei verschachtelten Klammern, wird der innerste Ausdruck immer zuerst ausgewertet
    * z.B. (a / (b - c))

#### 2. Potenzierungs-Operationen werden als nächstes ausgewertet
* Wenn ein Ausdruck mehrere Potenzierungsoperationen enthält, werden diese von rechts nach links ausgewertet (**rechtsassoziativ**)
    * a ** b ** c, ist dasselbe wie a ** (b ** c)

#### 3. Multiplikation, Division und Restwertberechnung werden als nächstes ausgewertet
* Wenn ein Ausdruck mehrere Multiplikationen, Division, Ganzzahl-Division und Modulo-Operationen enthält, wendet Python diese von links nach rechts an (**linksassoziativ**)
* Multiplikation, Division und Modulo haben die **gleiche Priorität**.

#### 4. Addition und Subtraktion werden zuletzt ausgewertet
* Wenn ein Ausdruck mehrere Additions- und Subtraktionsoperationen enthält, wendet Python diese von links nach rechts an (**linksassoziativ**). 
* Addition und Subtraktion haben die **gleiche Priorität**.

Wir werden diese Regeln erweitern, wenn wir andere Operationen einführen.


Komplette Liste: https://docs.python.org/3/reference/expressions.html#operator-precedence

## Gruppierung von Operationen mittels Klammern

In [19]:
1 + 2 + 3

6

In [20]:
(1 + 2) + 3

6

## Operandentypen haben Auswirkungen auf den Typ des Ergebnisses
Jeder arithmetische Oparator kann mit Ganzzahlen und Fliesskommazahlen verwendet werden.

Wenn beide Operanden ganze Zahlen sind: Ergebnis ist eine ganze Zahl - mit Ausnahme der Division (/), die immer eine Fliesskommazahl ergibt.

In [21]:
2 * 3

6

In [22]:
6 / 2

3.0

In [23]:
6 // 2

3

Wenn beide Operanden Fliesskommazahlen sind: Ergebnis ist eine Fliesskommazahl

In [24]:
3.0 * 2.0

6.0

Ausdrücke gemischter Art, die eine ganze Zahl und eine Fliesskommazahl enthalten, erzeugen immer ein Fliesskommaergebnis.

In [25]:
3 * 2.0

6.0